### PR TITLE
#1144 started GitLab integration

### PR DESCRIPTION
**Pull request by krzyk**
#1144 
Initial support for GitLab, created dummy webhook and added todos to do more implementation.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1171](https://github.com/zerocracy/farm/issues/1171) is now in scope, role is `REV`

File: src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ * GitLab hook, take.
+ *
+ * @author Krzysztof Krason (Krzysztof.Krason@gmail.com)
+ * @version $Id$
+ * @since 0.23
+ * @todo #1144:30min Create an implementation of GitLab API that will mirror
+ *  the one we use for Github (Github class and friends). We only need those
+ *  API endpoints that we use in farm project so don't try to implement the
+ *  full API.
+ * @todo #1144:30min Create a generic interface for hosted version control
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java]: @krzyk this task seems out of scope here. I think it should be a separate issue altogether. @g4s8 you might wanna weigh in here?

File: src/test/java/com/zerocracy/radars/gitlab/TkGitlabTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * GitLab hook, take.
+ *
+ * @author Krzysztof Krason (Krzysztof.Krason@gmail.com)
+ * @version $Id$
+ * @since 0.23
+ * @checkstyle JavadocMethodCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public final class TkGitlabTest {
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/radars/gitlab/TkGitlabTest.java]: @krzyk let's add a test for malformed (non-empty) JSON as well. Should be fairly straightforward.

File: src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    public Response act(final Request req) throws IOException {
+        if (
+            TkGitlab.json(
+                new InputStreamReader(req.body(), StandardCharsets.UTF_8)
+            ).isEmpty()
+        ) {
+            throw new RsForward(
+                new RsParFlash(
+                    new Par(
+                        "We expect this URL to be called by GitLub",
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java]: @krzyk I think you mean "Gitl**a**b" here.

**Comment by krzyk** [comment]
@carlosmiranda fixed (except the todo)

**Comment by carlosmiranda** [comment]
@krzyk all right, let's wait for comment from @g4s8

File: src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java (hunk 1 above)
Comment (g4s8) [reply src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java]: @carlosmiranda I agree, we need a task to implement some generic mechanism to work with jobs in stakeholders (and other classes), to distinguish GitHub and GitLab tickets, e.g. `gh:user/repo#1` and `gl:user/repo#1`. @krzyk I'd update this puzzle to clarify it

Comment (krzyk) [reply src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java]: @g4s8 Hmm, so @carlosmiranda said this shouldn't be a puzzle but a separate issue, and you say that I should update the puzzle description :)

So this puzzle should also mention that a stakeholder should be able to distinguish between github and gitlab (besides a generic interface for working on the repo), right?

**Comment by krzyk** [comment]
@carlosmiranda @g4s8 I've updated the todo description

File: src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java (hunk 1 above)
Comment (g4s8) [reply src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java]: @krzyk if this task is about different jobs, then it's not fully clear and I'd update it, but not create separate issue, because it was in origin task https://github.com/zerocracy/farm/issues/1144#issue-331923050 therefore it is uncompleted task (todo puzzle)

Comment (krzyk) [reply src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java]: @g4s8 yeah, so I updated the description. Is it more clear now?

Comment (g4s8) [reply src/main/java/com/zerocracy/radars/gitlab/TkGitlab.java]: @krzyk yes ,thanks

**Comment by krzyk** [comment]
@carlosmiranda I think we are good to go now (see comments above)

File: src/main/java/com/zerocracy/entry/Main.java [hunk 4]
```diff
@@ -133,6 +135,10 @@ public void exec() throws IOException {
                     new FkRegex(
                         "/ghook",
                         new TkMethods(new TkGithub(farm), "POST")
+                    ),
+                    new FkRegex(
+                        "/glhook",
+                        new TkMethods(new TkGitlab(), "POST")
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/entry/Main.java]: @krzyk can we use `HttpMethod.POST` here (same above) to avoid PMD suppression?

**Comment by krzyk** [comment]
@carlosmiranda updated

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1171](https://github.com/zerocracy/farm/issues/1171) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)
